File: src/main/java/org/noxet/noxetserver/minigames/MiniGameOptions.java
package org.noxet.noxetserver.minigames;

import org.bukkit.GameMode;

public interface MiniGameOptions {
    enum SpectatorContract {
        ALL,
        ONLY_DEAD_PLAYERS
    }

    String getId();
    String getDisplayName();

    int getMinPlayers();
    int getMaxPlayers();

    boolean allowPlayerDropIns();

    GameMode getDefaultGameMode();

    SpectatorContract getSpectatorContract();

    int getWorldChunksSquared();

    MiniGameController initGame();
}

File: src/main/java/org/noxet/noxetserver/minigames/MiniGameOptions.scala
package org.noxet.noxetserver.minigames

import org.bukkit.GameMode

trait MiniGameOptions:
    def getId: String
    def getDisplayName: String

    def getMinPlayers: Int
    def getMaxPlayers: Int

    /**
     * @return `true` if the game allows players to join when the game has started ("mid-game"), otherwise `false`.
     */
    def allowPlayerDropIns: Boolean

    def getDefaultGameMode: GameMode

    def getSpectatorContract: MiniGameOptions.SpectatorContract

    def getWorldChunksSquared: Int

    def initGame: MiniGameController

    def shouldAnnounceAdvancements: Boolean

object MiniGameOptions:
    enum SpectatorContract:
        case ALL, ONLY_DEAD_PLAYERS